### ISSUE TITLE
Add ignored byte feature

### ISSUE THREAD

**Issue by Evrey** [Alexhuszagh/rust-lexical#17]
Akin to the configurable exponent character etc., I'd like to be able to tell `lexical_core` that all `b'_'` are to be ignored.

I'm using this crate to parse floating-point literals in a toy programming language that allows arbitrary `_` separators to be added after the first digits before and after the `.`. Currently, I have to allocate memory just to throw away the `_` from otherwise valid input.

Test case:

`b"+4_2_.3_4_e+7_7_"` should successfully parse as `+42.34e77` if the to-be-ignored byte is set to `b'_'`.

Prior art:

- Rust's float literals allow for `_` separators.
- Ruby's as well.
- In C++14, `'` is a valid digit separator.

Questions:

- Are valid digits allowed? Who would even do that?
- Dot should not be allowed, I'd say?
- How would one treat the case of not having a to-be-ignored byte? Would that noticeably slow things down?
- Would it make sense to allow having multiple to-be-ignored bytes? Maybe even going so far as to add a `filter` callback? That'll for sure be slower.

**Comment by devsnek** [comment]
ideally you want to prevent repeats (`1__0`), trailing (`1_`, `1.3_`), and at decimal boundaries (`1_.3`, `1._3`), not just ignore the character entirely.

**Comment by Evrey** [comment]
@devsnek One could do that, yes. I specifically left those restrictions out, because in my case, the string already comes from a lexer enforcing some rules. It might be much easier and faster for `lexical` to just ignore such separators than to skip them while checking whether or not they follow some syntactic rules.

**Comment by Alexhuszagh** [comment]
@devsnek @Evrey Sorry for the delay (I've had some major mental health issues recently due to a bad reaction to medication, I'm back to normal), I like the idea of this feature. I'll look towards drafting an optional feature that integrates this in the future (optional, since obviously, it would not produce a standard-conforming float).

**Comment by Evrey** [comment]
@Alexhuszagh Health comes first. =)

Good to hear you recovered, and good to hear you like it.

**Comment by Alexhuszagh** [comment]
@Evrey So I'm thinking about the way to efficiently parse this, and ideally, we'd want it to be consistent with how the Rust parser accepts floating-point literals, or at least other popular programming languages. Using Rust, the most lenient for the grammar, actually simplifies the implementation a lot.

Here are some basic separated float lliterals in different programming languages, with commented-out values means they're invalid:

**Rust**

```rust
3.0_1f32
3.0__1f32
//3._01f32
3_.01f32
3.01_f32
//3.01f_32
3.01_e1_f32
```

**Python**

```python
3.0_1
#3.0__1
#3._01
#3_.01
#3.01_
#3.01_e1
```

**Ruby**

```ruby
3.0_1
#3.0__1
#3._01
#3_.01
#3.01_
#3.01_e1
```

In short, there are two predominant modes, the Ruby/Python way, where `_`s must be internal to the number, they cannot be adjacent to the `.` or exponent notation character, and cannot be adjacent to each other. In Rust, the implementation is much simpler: you cannot have a leading `_`  or an `_` after a `.`, for naming issues: anything else goes.

I can reasonably implement both, the simplest would be the Rust variant. Which would you prefer first, and what feature names would be ideal for this?

**Comment by Alexhuszagh** [comment]
For feature names, I've contemplated `tidy_floats`, but we need some way to distinguish the Python/Ruby and Rust syntax. Any suggestions would be wonderful. I'll likely implement Rust's syntax first, since it's very easy to implement internally.

**Comment by Alexhuszagh** [comment]
Also, another question, should we configure the ignored byte to be configurable, as the exponent characters, infinity/NaN strings currently are, or should it be hard-coded as `_`.

- [ Feature Naming (`tidy_floats`?) ] 
- [ Syntax (Rust or Python or both) ]
- [ Do we hardcode `_` or allow the character to be configured? ]

**Comment by Alexhuszagh** [comment]
As far as having valid digits or control characters for a float being allowed as ignored characters, I'd say we should strictly disallow them. Therefore, invalid characters should always be: `[0-9._+-]`, and the exponent character(s).

I'd consider supporting a configurable character to ignore, however, the only reason C++ uses `'` is because of a mess of backwards compatibility (specifically, literal suffixes like _1). Is this a valid-enough use-case?

Specifically, my answers to the above questions are:

1. `tidy_floats_rust` and `tidy_floats_python`
2. Both
3. Configurable, disallow certain control characters.

**Comment by Alexhuszagh** [comment]
As far as performance goes, the Python variant may have some noticeable performance hits since it requires you to track the previous character and ensure the last character is not an ignored character. The Rust variant should have effectively no performance penalty, since it can be implement in a match statement (which can get optimized very effectively).

```rust
match &c {
   b'0' ... b'9' => do_something(c),
   b'_' => ignore(c),
}
```

Even if it's not hard-coded into a jump table via a configurable ignore character, the latter option should have minimal performance overhead.

**Comment by Evrey** [comment]
> But we need some way to distinguish the Python/Ruby and Rust syntax
>
> some noticeable performance hits since it requires you to track the previous character

Honestly, I think this goes way too far. I think it is not at all `lexical`'s job to decide what valid digit separation syntax is. Like in my case, the float literal I feed into `lexical` comes right out of a `logos` token. I already verified what I think valid separators are. If `lexical` does that very same job twice now, it just ends up wasting precious CPU time. Worst case, `lexical` may be incompatible to some other programming language that has yet another separator style. And I haven't yet found production code that feeds some arbitrary text into some float parser before checking the token syntax.

Hence why I suggested this feature as "ignored byte". Treat them like line noise. This also keeps `lexical`s code fast. It does not have to bother with any relative positioning to other bytes. Just skip and go on with the actual float parsing. It may even be that slow downs due to skip byte handling are so negligible that it doesn't even need to be feature-gated.

> In short, there are two predominant modes
>
> Also, another question, should we configure the ignored byte to be configurable, as the exponent characters

I think the more interesting observation here is that no useful programming language out there allows digit separators that overlap with `[0-9] | '-' | '+' | [a-zA-Z] | '.'` and only picks one separator char. And that all programming languages I know pick a separator out of basic ASCII. So configuring a single skip byte that does not match these characters does the job.

> the only reason C++ uses `'` is because of a mess of backwards compatibility

That, and in some countries that's the common digit separator.

**Comment by Evrey** [comment]
Addendum to the skip byte. If the performance impact turns out to be indeed not worth noting, then the byte `0xC1` may be used as a default ignored byte. `0xC1` is an always-invalid UTF-8 byte and an always-invalid ASCII byte, so it'll try filtering a byte that will never come. Potentially cheaper in runtime cost and definitely cheaper in storage cost than `Option`-wrapping the byte.

**Comment by Alexhuszagh** [comment]
I know it's been a long time, but I've finally started making significant progress on this. I've been considering both a comprehensive solution and a cheap solution, and luckily one can be implemented in terms of the other. So far in my benchmarks it doesn't seem to be slowing performance either, although we'll see later. Everything is fairly preliminary, but so far is going quite well.

The tree [float_format](https://github.com/Alexhuszagh/rust-lexical/tree/float_format) is tracking changes for this.

**Comment by Evrey** [comment]
@Alexhuszagh Glad yer back. °^° I'm patient, I can wait.

Edit: Ah, damn clicked the wrong button! D:<

**Comment by Alexhuszagh** [comment]
@Evrey This feature is now live on the `float_formats` branch, however, the last stage added a bit of a performance penalty so I will be debugging that cause before merging it into master.

**Comment by Alexhuszagh** [comment]
Ok, these changes have been merged into master for both the v0.6 and v0.7 branches (v0.7 is master, and supports Rustc 1.37.0+, and v0.6 is the maintenance branch, which supports Rustc 1.24.0+).

The new syntax is as follows:

```rust
extern crate lexical_core;  // ^0.6.4 or ^0.7.1

pub fn main() {
     // Use the pre-defined JSON format.
    let format = lexical_core::NumberFormat::JSON;
    println!("{:?}", lexical_core::parse_format("3.e7", format));    // Err

    // Let's use digit separators, ignore them, and have a permissive float grammar.
    // This only throws if the digit separator isn't valid ASCII, so you can safely unwrap.
    let format = lexical_core::NumberFormat::ignore(b'_').unwrap();
    println!("{:?}", lexical_core::parse_format("1_2.3_4e+6_7", format));    // Ok(12.34e67)
}
```

**Comment by Evrey** [comment]
This looks really nice, great work! I like the idea of `NumberFormat::JSON` and all the others. Really, that's a ton of predefined settings you made there!

Btw., just noticed this: https://github.com/Alexhuszagh/rust-lexical/blob/11ad6646c183d39a743a3f6d36b62d8c84ed4e86/lexical-core/src/util/format.rs#L789 Did you mean »required mantissa positive sign«?

**Comment by Alexhuszagh** [comment]
@Evrey Yes, yes I did thank you.

**Comment by Alexhuszagh** [comment]
This feature is now live in crates.io. Thanks for all the feedback and the suggestion to implement this.
